examples/interpreter-demo/serialize.js

```diff
@@ -0,0 +1,411 @@
+      interpreter.ast.loc.end.constructor;
+}
+
+function deserialize(json, interpreter) {
+  function decodeValue(value) {
+    if (value && typeof value === 'object') {
+      var data;
+      if ((data = value['#'])) {
+       // Object reference: {'#': 42}
+       value = objectList[data];
+        if (!value) {
+          throw ReferenceError('Object reference not found: ' + data);
+        }
+        return value;
+      }
+      if ((data = value['Number'])) {
+        // Special number: {'Number': 'Infinity'}
+        return Number(data);
+      }
+      if ((data = value['Value'])) {
+        // Special value: {'Value': 'undefined'}
+        if (value['Value'] === 'undefined') {
+          return undefined;
+        }
+      }
+    }
+    return value;
+  }
+  var stack = interpreter.getStateStack();
+  if (!Array.isArray(json)) {
+    throw TypeError('Top-level JSON is not a list.');
+  }
+  if (!stack.length) {
+    // Require native functions to be present.
+    throw Error('Interpreter must be initialized prior to deserialization.');
+  }
+  recordAcornConstructons(interpreter);
+  var LOC_REGEX = /^(\d*):(\d*)-(\d*):(\d*) ?(.*)$/;
+  // Find all native functions in existing interpreter.
+  var objectList = [];
+  objectHunt_(stack, objectList);
+  var functionHash = Object.create(null);
+  for (var i = 0; i < objectList.length; i++) {
+    if (typeof objectList[i] === 'function') {
+      functionHash[objectList[i].id] = objectList[i];
+    }
+  }
+  // First pass: Create object stubs for every object.
+  objectList = [];
+  for (var i = 0; i < json.length; i++) {
+    var jsonObj = json[i];
+    var obj;
+    switch (jsonObj['type']) {
+      case 'Map':
+        obj = Object.create(null);
+        break;
+      case 'Object':
+        obj = {};
+        break;
+      case 'ScopeReference':
+        obj = Interpreter.SCOPE_REFERENCE;
+        break;
+      case 'Function':
+        obj = functionHash[jsonObj['id']];
+        if (!obj) {
+          throw RangeError('Function ID not found: ' + jsonObj['id']);
+        }
+        break;
+      case 'Array':
+        // Currently we assume that Arrays are not sparse.
+        obj = [];
+        break;
+      case 'Date':
+        obj = new Date(jsonObj['data']);
+        if (isNaN(obj)) {
+          throw TypeError('Invalid date: ' + jsonObj['data']);
+        }
+        break;
+      case 'RegExp':
+        obj = RegExp(jsonObj['source'], jsonObj['flags']);
+        break;
+      case 'PseudoObject':
+        obj = new Interpreter.Object(null);
+        break;
+      case 'Scope':
+        obj = new Interpreter.Scope(undefined, undefined, undefined);
+        break;
+      case 'State':
+        obj = new Interpreter.State(undefined, undefined);
+        break;
+      case 'Node':
+        obj = new NODE_CONSTRUCTOR();
+        delete obj.start;
+        delete obj.end;
+        var locText = jsonObj['loc'];
+        if (locText) {
+          var loc = new NODE_LOC_CONSTRUCTOR();
```

Comment:
Break this out into a helper function and add comments, writing for a reader who doesn't know the structure of Nodes in JS-Interpreter.